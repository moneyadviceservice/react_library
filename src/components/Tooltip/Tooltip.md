Tooltip

```jsx
import { Tooltip, Row, Button, Paragraph as P } from '@moneypensionservice/directories';

const lorem = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.';

<Row justify="space-evenly" align="center" direction="row">
  <P width="auto" margin="0">
    Default tooltip <Tooltip text={`This is a default tooltip.`} />
  </P>
  <Tooltip text={lorem} hover side="top">
    <Button text="Hover me!" />
  </Tooltip>
  <Tooltip text={lorem} side="bottom" minWidth={{xs: '150px', md: '250px'}}>
    Click me!
  </Tooltip>
  <Tooltip text={lorem} side={{xs: 'bottom', md:'left'}}>
    xs: bottom, md:left
  </Tooltip>
</Row>
```
