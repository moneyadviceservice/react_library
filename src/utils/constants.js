export const breakpoints = {
  xs: 1,
  sm: 48, // 768px
  md: 64, // 1024px
  lg: 90, // 1440px
  xl: 120, // 1920px
}

export const dimensions = ['xs', 'sm', 'md', 'lg', 'xl']

export const mediaQuery = 'only screen'

export const ALIGN_SELF_MAP = {
  center: 'center',
  end: 'flex-end',
  start: 'flex-start',
  stretch: 'stretch',
}

export const SPACING_SIZES = [
  'xxsmall',
  'xsmall',
  'small',
  'medium',
  'large',
  'xlarge',
]

