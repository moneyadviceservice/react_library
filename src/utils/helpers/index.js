export * from './background'
export * from './responsive'
export * from './flexbox'
export * from './styles'
export * from './mediaQueries'
export * from './sizing'
export * from './refs'
