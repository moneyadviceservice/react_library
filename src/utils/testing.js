import { render, act } from '@testing-library/react'
import GlobalTheme from '../components/ThemeProvider'

const customRender = (ui, options) =>
  render(ui, { wrapper: GlobalTheme, ...options })

const waitRender = async (ui, options) => {
  const { container, ...rest } = customRender(ui, options)

  await act(async () => {
    await Promise.resolve(container)
  })

  return { container, ...rest }
}

export * from '@testing-library/react'

export { customRender as render, waitRender }
