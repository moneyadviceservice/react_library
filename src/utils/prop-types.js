import PropTypes from 'prop-types'
import { SPACING_SIZES } from './constants'

/** spacing prop.type */
const spacingProp = PropTypes.shape({
  bottom: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  horizontal: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  left: PropTypes.oneOfType([PropTypes.oneOf(SPACING_SIZES), PropTypes.string]),
  right: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  top: PropTypes.oneOfType([PropTypes.oneOf(SPACING_SIZES), PropTypes.string]),
  vertical: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
})

export const genericPropTypes = {
  a11yTitle: PropTypes.string,
  alignSelf: PropTypes.oneOf(['start', 'center', 'end', 'stretch']),
  margin: PropTypes.oneOfType([
    PropTypes.oneOf(['none', ...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
  padding: PropTypes.oneOfType([
    PropTypes.oneOf([...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
  border: PropTypes.oneOfType([
    PropTypes.oneOf(['none', ...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
}

export function genericPropsDefaults({
  a11yTitle = null,
  alignSelf = null,
  margin = null,
  padding = null,
  border = null,
} = {}) {
  return {
    a11yTitle: a11yTitle,
    alignSelf: alignSelf,
    margin: margin,
    padding: padding,
    border: border,
  }
}
