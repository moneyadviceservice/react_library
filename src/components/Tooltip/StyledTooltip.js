import styled, { css } from 'styled-components'
import { genericStyles, responsiveProps } from '../../utils/helpers'
import { Anchor } from '../Anchor'

const StyledTooltip = styled.span`
  ${genericStyles}
  position: relative;
  cursor: pointer;
`

/** Side styles for popup */
const topStyle = css`
  top: auto;
  right: auto;
  left: 50%;
  bottom: calc(100% + 5px);
  transform: translate(-50%, -0.5em);
`
const rightStyle = css`
  bottom: auto;
  top: 50%;
  left: calc(100% + 5px);
  right: calc(0em - 5px);
  transform: translate(0.5em, -50%);
`
const bottomStyle = css`
  right: auto;
  bottom: auto;
  left: 50%;
  top: calc(100% + 5px);
  transform: translate(-50%, 0.5em);
`
const leftStyle = css`
  bottom: auto;
  left: auto;
  top: 50%;
  right: calc(100% + 5px);
  transform: translate(-0.5em, -50%);
`

const sideStyles = {
  top: topStyle,
  right: rightStyle,
  bottom: bottomStyle,
  left: leftStyle,
}

const renderSide = sideProp => {
  // string - fixed width
  if (typeof sideProp === 'string') return sideStyles[sideProp]
  // responsive
  if (typeof sideProp === 'object') {
    const responsiveStyles = {}
    for (let [key, value] of Object.entries(sideProp)) {
      responsiveStyles[key] = sideStyles[value]
    }
    return responsiveProps(null, responsiveStyles)
  }
}

const Tip = styled.span`
  /** Popup pseudo element */
  &:after {
    /** conditional styles */
    display: ${({ show }) => (show ? 'block' : 'none')};
    ${({ minWidth }) => responsiveProps('min-width', minWidth)}
    /** positioning */
    position: absolute;
    z-index: 1000;
    /** Prop-based side styles */
    ${({ side }) => renderSide(side)}
    /** popup content */
    content: "${({ text }) => text}";
    padding: 8px;
    border: 1px solid black; 
    border-radius: 4px;
    font-size: 0.875rem;
    line-height: 1rem;
    white-space: pre-wrap;
    background-color: #FFF;
    box-shadow: 0 1em 2em -0.5em rgba(0, 0, 0, 0.35);
  }
`

const Icon = styled(Anchor)`
  font-style: italic;
  padding: 0.375rem;
`

export { StyledTooltip, Icon, Tip }
