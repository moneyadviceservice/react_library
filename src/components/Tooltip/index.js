import React, { useState, useRef } from 'react'
import PropTypes from 'prop-types'
import { genericPropTypes, genericPropsDefaults } from '../../utils/prop-types'
import { StyledTooltip, Icon, Tip } from './StyledTooltip'

const Tooltip = ({
  a11yTitle,
  children,
  hover,
  minWidth,
  side,
  text,
  ...rest
}) => {
  const node = useRef()
  const [show, setShow] = useState(false)

  const handleClick = () => {
    if (!show) {
      setShow(true)
      document.addEventListener('click', handleOutsideClick, false)
    } else {
      setShow(false)
      document.removeEventListener('click', handleOutsideClick, false)
    }
  }

  const handleOutsideClick = e => {
    if (node.current && !node.current.contains(e.target)) {
      setShow(false)
      document.addEventListener('click', handleOutsideClick, false)
    }
    return
  }

  return (
    <StyledTooltip
      aria-label={a11yTitle}
      onMouseEnter={() => hover && setShow(true)}
      onMouseLeave={() => hover && setShow(false)}
      ref={node}
      {...rest}
    >
      {children ? (
        <span onClick={() => !hover && handleClick()}>{children}</span>
      ) : (
        <Icon weight={700} onClick={() => !hover && handleClick()}>
          i
        </Icon>
      )}
      {show && (
        <Tip minWidth={minWidth} side={side}>
          {text}
        </Tip>
      )}
    </StyledTooltip>
  )
}

// Documentation
Tooltip.propTypes = {
  /** Content inside component. If children are available inside the tooltip component they'll work as the trigger. */
  children: PropTypes.node,
  /** If enabled, the tooltip will show on hover. */
  hover: PropTypes.bool,
  /** Sets the minimum width of the tooltip. Responsive.e */
  minWidth: PropTypes.oneOfType([PropTypes.object, PropTypes.string]),
  /** Defines the side where the tooltip shows. Responsive. */
  side: PropTypes.oneOfType([PropTypes.object, PropTypes.string]),
  /** The text to display inside the Tooltip. */
  text: PropTypes.any,
  ...genericPropTypes,
}

Tooltip.defaultProps = {
  minWidth: '160px',
  side: 'right',
  ...genericPropsDefaults(),
}

/** @component */
export { Tooltip }
