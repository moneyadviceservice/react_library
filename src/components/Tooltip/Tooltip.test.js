import React from 'react'
import { render, fireEvent, act } from '../../utils/testing'
import { Tooltip } from '../Tooltip'

describe('<Tooltip />', () => {
  describe('Snapshot', () => {
    it('should match click snapshot', () => {
      const { container } = render(<Tooltip text="Tooltip text" />)
      // tooltip closed
      expect(container).toMatchSnapshot()
      // click on icon
      act(() => {
        fireEvent.click(container.firstChild)
      })
      // tooltip opened
      expect(container).toMatchSnapshot()
    })

    it('should match hover snapshot with children', () => {
      const { container } = render(
        <Tooltip hover text="Tooltip text">
          <span>hover</span>
        </Tooltip>
      )
      // tooltip closed
      expect(container).toMatchSnapshot()
      // hover
      act(() => {
        fireEvent.mouseEnter(container)
      })
      // tooltip opened
      expect(container).toMatchSnapshot()
    })
  })
})
