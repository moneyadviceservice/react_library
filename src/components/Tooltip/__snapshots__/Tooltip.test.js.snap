// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tooltip /> Snapshot should match click snapshot 1`] = `
<div>
  <span
    class="StyledTooltip-sc-1uvk9zg-0 dWzGPd"
  >
    <a
      class="StyledParagraph-sc-1toaej-0 drRbGa StyledAnchor-sc-1qj1ci5-0 fNhjgf StyledTooltip__Icon-sc-1uvk9zg-2 kLqhMp"
    >
      i
    </a>
  </span>
</div>
`;

exports[`<Tooltip /> Snapshot should match click snapshot 2`] = `
<div>
  <span
    class="StyledTooltip-sc-1uvk9zg-0 dWzGPd"
  >
    <a
      class="StyledParagraph-sc-1toaej-0 drRbGa StyledAnchor-sc-1qj1ci5-0 fNhjgf StyledTooltip__Icon-sc-1uvk9zg-2 kLqhMp"
    >
      i
    </a>
  </span>
</div>
`;

exports[`<Tooltip /> Snapshot should match hover snapshot with children 1`] = `
<div>
  <span
    class="StyledTooltip-sc-1uvk9zg-0 dWzGPd"
  >
    <span>
      <span>
        hover
      </span>
    </span>
  </span>
</div>
`;

exports[`<Tooltip /> Snapshot should match hover snapshot with children 2`] = `
<div>
  <span
    class="StyledTooltip-sc-1uvk9zg-0 dWzGPd"
  >
    <span>
      <span>
        hover
      </span>
    </span>
  </span>
</div>
`;
